anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java

```diff
@@ -267,75 +264,37 @@ public FetchResponseCollector(K key) {
-
-      public BackingEntrySet(CacheSet<CacheEntry<K, V>> entrySet) {
-         super(cache.wired(), entrySet);
-         this.entrySet = entrySet;
-      }
-
-      @Override
-      public CloseableIterator<CacheEntry<K, V>> iterator() {
-         // It's safe to use entrySet.iterator() because everything happens on the originator
-         return new BackingIterator(entrySet.iterator());
-      }
-
-      @Override
-      public CloseableSpliterator<CacheEntry<K, V>> spliterator() {
-         return Closeables.spliterator(iterator(), Long.MAX_VALUE,
-                                       Spliterator.CONCURRENT | Spliterator.DISTINCT | Spliterator.NONNULL);
-      }
-   }
-
-   private class BackingIterator implements CloseableIterator<CacheEntry<K, V>> {
-
-      private Iterator<CacheEntry<K, V>> iterator;
+   private class BackingEntrySet extends InternalCacheSet<CacheEntry<K, V>> {
+      protected final InternalCacheSet<CacheEntry<K, V>> next;
 
-      public BackingIterator(Iterator<CacheEntry<K, V>> iterator) {
-         this.iterator = iterator;
+      public BackingEntrySet(InternalCacheSet<CacheEntry<K, V>> next) {
+         this.next = next;
       }
 
       @Override
-      public void close() {
-         if (iterator instanceof CloseableIterator<?>) {
-            ((CloseableIterator<?>) iterator).close();
-         }
+      public Publisher<CacheEntry<K, V>> localPublisher(int segment) {
+         return Flowable.fromPublisher(next.localPublisher(segment))
+                        .map(localEntry -> fixPublisherEntry(segment, localEntry));
       }
 
       @Override
-      public boolean hasNext() {
-         if (iterator == null) {
-            return false;
-         }
-         return iterator.hasNext();
+      public Publisher<CacheEntry<K, V>> localPublisher(IntSet segments) {
+         return Flowable.fromPublisher(next.localPublisher(segments))
+                        .map(localEntry -> {
+                           int segment = distributionManager.getCacheTopology().getSegment(localEntry.getKey());
+                           return fixPublisherEntry(segment, localEntry);
+                        });
       }
 
-      @Override
-      public CacheEntry<K, V> next() {
-         if (iterator == null) {
-            throw new NoSuchElementException();
-         }
-         CacheEntry<K, V> localEntry = iterator.next();
+      private CacheEntry<K, V> fixPublisherEntry(int segment, CacheEntry<K, V> localEntry) {
          if (localEntry.getMetadata() instanceof RemoteMetadata) {
             // The key exists and the anchor location is a remote node
             RemoteMetadata remoteMetadata = (RemoteMetadata) localEntry.getMetadata();
             Address keyLocation = remoteMetadata.getAddress();
 
-            int segment = distributionManager.getCacheTopology().getSegment(localEntry.getKey());
             return CompletionStages.join(getRemoteValue(keyLocation, localEntry.getKey(), segment, false));
```

Comment:
I'm not sure I want to do that, because then I'd have to also worry about what's a good number of concurrent requests -- I don't trust flatMap's defaults ;)

Thanks for noticing the code is blocking though, looks like I moved from iterator to publisher without actually reading the implementation!

Reply:
Well it isn't necessarily faster, but the code don't have to check the segment or allocate a Maybe for each of the local elements.

Either way is fine though, I don't think it is a big deal. But it does need to use `concatMap` at a minimum to handle the blocking part.

Reply:
Yep, I did it with flatMap in the end.